### DISCUSSION TITLE
Axum Post method trait bound error

### DISCUSSION THREAD

**Discussion by juddbaguio** [tokio-rs/axum#2239] [Q&A]
- [x] I have looked for existing issues (including closed) about this

### Crates:
├── axum v0.6.20
│   ├── axum-core v0.3.4
│   ├── axum-macros v0.3.8 (proc-macro)

### Description: Post Handler Trait Bound Error

I have come across of an Axum trait bound handler error:
```rust
error[E0277]: the trait bound `fn(axum::extract::State<std::sync::Arc<services::user_service::Context>>, axum::Json<domain::dto::user_dto::CreateUserDTO>) -> impl std::future::Future<Output = std::result::Result<axum::Json<domain::entities::User>, (axum::http::StatusCode, std::string::String)>> {api::users::create_user::handler}: axum::handler::Handler<_, _, _>` is not satisfied
   --> src\api\users\mod.rs:11:26
    |
11  |         .route("/", post(create_user::handler))
    |                     ---- ^^^^^^^^^^^^^^^^^^^^ the trait `axum::handler::Handler<_, _, _>` is not implemented for fn item `fn(axum::extract::State<std::sync::Arc<services::user_service::Context>>, axum::Json<domain::dto::user_dto::CreateUserDTO>) -> impl std::future::Future<Output = std::result::Result<axum::Json<domain::entities::User>, (axum::http::StatusCode, std::string::String)>> {api::users::create_user::handler}`
    |                     |
    |                     required by a bound introduced by this call
    |
    = help: the following other types implement trait `axum::handler::Handler<T, S, B>`:
              <axum::handler::Layered<L, H, T, S, B, B2> as axum::handler::Handler<T, S, B2>>
              <axum::routing::MethodRouter<S, B> as axum::handler::Handler<(), S, B>>
```

And this is my handler implementation:
```rust
#[debug_handler]
pub async fn handler(
    State(user_service): State<Arc<Context>>,
    Json(payload): Json<CreateUserDTO>,
) -> Result<Json<User>, (StatusCode, String)> {
    let res = user_service.create_user(payload).await;
    if res.is_err() {
        return Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            res.err().unwrap().to_string(),
        ));
    }

    Ok(Json(res.unwrap()))
}
```

I am not sure where it went wrong since I correctly follow the order of extractors. Any guidance or solutions will be appreciated, thank you!

Edit:
More info on the `Context` struct:
```rust
#[derive(Clone)]
pub struct Context {
    manager: TransactionManager,
}

pub trait ThreadSafe: Send + Sync {}

impl ThreadSafe for Context {}

#[derive(Clone)]
pub struct TransactionManager {
    pg_pool: Pool<Postgres>,
}

impl ThreadSafe for TransactionManager {}
```

Edit: Service function (create_user) implementation, I am using launchbadge/sqlx as my database driver. If I resort to `&Pool<Postgres>` connection type, the error seems gone. This only happens when I try to run transactions,
![image](https://github.com/tokio-rs/axum/assets/52206958/77db292b-238d-43a0-b8da-4a12df83b9c2)

**Comment by davidpdrsn** [comment]
Did you try https://docs.rs/axum/0.6.20/axum/attr.debug_handler.html?

**Reply by juddbaguio** [reply]
I already did and the debug handler didn't show some kind of errors.

**Reply by juddbaguio** [reply]
No errors were shown when I added the debug_handler macro
![image](https://github.com/tokio-rs/axum/assets/52206958/08271d7c-7079-4596-9610-cf94c4b02cae)

**Reply by juddbaguio** [reply]
hello @davidpdrsn , I added more info from my second investigation.

**Comment by juddbaguio** [comment]
I am going to close this discussion, considered as Resolve.

The issue found is in my `users_repo::create_user` which takes `Acquire` trait from sqlx and I have found that this trait is not implemented for `Send`.
![image](https://github.com/tokio-rs/axum/assets/52206958/0a2a75f5-60a0-4010-b165-e4b3d7f24af8)

I modified the db executor to be on `Executor<'_,DB>` trait which has supertrait of `Send`
![image](https://github.com/tokio-rs/axum/assets/52206958/f41f461a-c281-4b65-a4e4-3db3fb03cec6)

**Reply by trchopan** [reply]
Hi @juddbaguio ,
I have the same issue using sqlite and axum. Do you mind explain more how did you resolve your issue?
Thank you.

**Reply by juddbaguio** [reply]
Hi, basically Axum handlers required all objects to have a 'Send' trait. You may need to check all your functions involved specifically from your db structs or anything else if they already auto implemented the 'Sync' trait.

**Reply by trchopan** [reply]
Thank you for explain.

For me, it may not the same issue as you but the order of the extractor was the cause.

When I try
![image](https://github.com/tokio-rs/axum/assets/4874258/4601ecc5-eb9d-4185-8c8c-e3b9ded3d44e)

I got this error
![image](https://github.com/tokio-rs/axum/assets/4874258/f372d4f1-1321-4788-8569-3047163d8140)

But if I put the `extract::Json` extractor at the end. It is compiled and works fine.
![image](https://github.com/tokio-rs/axum/assets/4874258/fd8a9b99-8b1c-4bbe-a7aa-a4f5cb81edff)
